modules/substrate/src/verifier.rs

```diff
@@ -924,4 +924,89 @@ mod tests {
 			assert_ok!(verifier.import_header(child.clone()));
 		})
 	}
+
+	#[ignore]
+	#[test]
+	fn checks_if_nando_knows_what_were_supposed_to_handle() {
+		run_test(|| {
+			// Legend:
+			//    S|N: Schedules change in N blocks
+			//    E: Enacts change
+			//    F: Finalized
+			//    Order: Import order
+			//      FN: Finality proof imported for header N
+			//
+			//  Note: We assume 1 is `last_finalized` at start
+			// ---
+			//
+			// [1] <- [2: F]
+			//   \ [2']
+			//
+			// Order: 1, 2, 2', F2, F2'
+			//
+			// Not allowed to finalize 2'
+			// Can't import finality proof since not ancestor of 2
+			//
+			// ---
+			//
+			//   / [2: S|1] <- [3]
```

Comment:
Iirc if change is scheduled in 1 block, then it'll be enacted at 3 => block 3 requires justification. I think you mean to have `S|0` here instead. Then everything looks fine.